### PR TITLE
refs #126151 [Admin] Login

### PR DESCRIPTION
**Pull request by nghia2742-briswell**

### CODE REVIEW

File: app/Http/Controllers/AuthController.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+        $credentials = $request->only('email', 'password');
+        $email = $credentials['email'];
+        $password = $credentials['password'];
+
+        if (Auth::attempt(['email' => $email, 'password' => $password, 'user_flg' => 0])) {
+            return redirect()->intended('admin');
+        }
+
+        $yamlPath = file_get_contents('../messages.yaml');
```
Comment (bwv-hieu-nt) [inline app/Http/Controllers/AuthController.php:35]: $yamlPath = file_get_contents('../messages.yaml');
$yamlContents = Yaml::parse($yamlPath);

Tạo 1 thư mục utils, viết 1 class và function để tái xử dụng lại ở đoạn code trên
VD: app/Utils/MessageUtils.php. Cân nhắc việc sử dụng static function hoặc instance class 
class MessageUtil
public static function getMesssage(x1,x2)
Usage: MessageUtil::getMessage(x1,x2)

File: messages.yaml
Comment (bwv-hieu-nt) [inline messages.yaml:1]: File này cần đem vào thư mục nào đó để dễ quản lý các file constants, vì sau này sẽ còn tạo nhiều file constant khác như common.yml, [table].yml
VD: 
config/constants/messages.yml
config/constants/common.yml
config/constants/user.yml
......

File: resources/views/layouts/master.blade.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE html>
+<html lang="{{ str_replace('_', '-', app()->getLocale()) }}">
+    <head>
+        <meta charset="utf-8">
+        <meta name="viewport" content="width=device-width, initial-scale=1">
+
+        <title>@yield('title', config('app.name', '@Master Layout'))</title>
+
+        {{-- BS5 --}}
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-EVSTQN3/azprG1Anm3QDgpJLIm9Nao0Yz1ztcQTwFspd3yD65VohhpuuCOmLASjC" crossorigin="anonymous">
```
Comment (bwv-hieu-nt) [inline resources/views/layouts/master.blade.php:10]: Nên tải xuống các thư viện file css/js, không dùng CDN
Các file tĩnh như thư viện boostrap có thể đặt tại thư mục public thay vì resources
Lưu ý: nên lấy các file .min.css để hạn chế mặc loading file từ client

File: resources/views/layouts/master.blade.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        @vite(['resources/css/app.css'])
+    </head>
+    <body>
+    
+        @yield('content')
+
+        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
```
Comment (bwv-hieu-nt) [inline resources/views/layouts/master.blade.php:17]: Nên tải xuống các thư viện file css/js, không dùng CDN

File: app/Http/Controllers/AuthController.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            Session::put('user', $user);
+        }
+        return view('index');
+    }
+
+    public function login() {
+        return view('auth.login');
+    }
+
+    public function handleLogin(LoginRequest $request) {
```
Comment (bwv-hieu-nt) [inline app/Http/Controllers/AuthController.php:25]: các function cần comment 
sử dụng cú pháp của vscode
/** + enter

File: app/Http/Controllers/AuthController.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+use App\Http\Requests\LoginRequest;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Session;
+use Symfony\Component\Yaml\Yaml;
+
+class AuthController extends Controller
+{
+    public function index()
+    {
```
Comment (bwv-hieu-nt) [inline app/Http/Controllers/AuthController.php:13]: Thống nhất 1 style
public function index () {

}

File: resources/js/main.js
Comment (bwv-hieu-nt) [inline resources/js/main.js:1]: main.js 
Sẽ hiểu nhầm tên file dùng chưa đúng chủ đích của màn hình login
Cần sửa lại và đặt gọn vào các thư mục để dễ quản lý
VD:
js/screens/login.js

File: resources/views/auth/login.blade.php
Comment (bwv-hieu-nt) [inline resources/views/auth/login.blade.php:1]: views/auth/login.blade.php
Sau này sẽ nhiều screens cần tạo 1 thư mục cha để quản lý
VD: views/screens/auth/login.blade.php
views/screens/user/...
views/screens/product/....

File: resources/views/welcome.blade.php
Comment (bwv-hieu-nt) [inline resources/views/welcome.blade.php:1]: Xóa file này nếu không cần dùng đến

File: routes/web.php [hunk 6]
```diff
@@ -14,5 +15,23 @@
 */
 
 Route::get('/', function () {
```
Comment (bwv-hieu-nt) [inline routes/web.php:17]: Xóa code không sử dụng

File: app/Http/Controllers/AuthController.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        return view('auth.login');
+    }
+
+    public function handleLogin(LoginRequest $request) {
+
+        $credentials = $request->only('email', 'password');
+        $email = $credentials['email'];
+        $password = $credentials['password'];
+
+        if (Auth::attempt(['email' => $email, 'password' => $password, 'user_flg' => 0])) {
```
Comment (bwv-hieu-nt) [inline app/Http/Controllers/AuthController.php:31]: 'user_flg' => 0
Không nên gán mặc định các giá trị như 0,1,2,3.... Vì sau này nếu user_flg thay đổi điều kiện thì sẽ sửa rất nhiều chỗ đang viết kiểu này
Nên tạo 1 common.yml và tạo các value list cho nó và gọi ra sử dụng
VD:
user_flg:
admin: 0
user: 1

File: app/Http/Controllers/AuthController.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+        $yamlPath = file_get_contents('../messages.yaml');
+        $yamlContents = Yaml::parse($yamlPath);
+        $errorMsg = $yamlContents['errors']['E010'];
+        
+        return redirect()->back()->withInput(['email' => $email, 'password' => $password])->with('errorMsg', $errorMsg);
+    }
+
+    public function logout() {
+        Auth::logout();
```
Comment (bwv-hieu-nt) [inline app/Http/Controllers/AuthController.php:43]: logout cần thiết xóa hết các session đang khởi tạo

File: resources/js/main.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+                required: true,
+                email: true,
+            },
+            password: {
+                required: true,
+            },
+        },
+        messages: {
+            email: {
+                required: "Email is required field.",
```
Comment (bwv-hieu-nt) [inline resources/js/main.js:14]: Không set thẳng các message như thế này, vì sau này nếu có thay đổi sẽ sửa nhiều nơi
Cần tạo 1 file common.js, để gọi các message về truyền vào
JS: thao khảo jquery extend

File: resources/js/main.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+                email: "Please enter your email address correctly.",
+            },  
+            password: {
+                required: "Password is required field.",
+            },   
+        },
+        invalidHandler: function(form, validator) {
```
Comment (bwv-hieu-nt) [inline resources/js/main.js:21]: Cần tạo 1 file jquery validate common để xử lý chung các việc focusout, handle error message
Chưa xử lý 1 lần nhấn button. Tham khảo submitHandler và xử lý khi nhấn button submit sẽ show 1 hình loading (CSS loader and spinner)

File: resources/js/main.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+            var errors = validator.numberOfInvalids();
+            if (errors) {                    
+                validator.errorList[0].element.focus();
+            }
+        } 
+    });
+})
+
+$('#loginForm').submit(function (e) {
+    if ($('#email').val() !== '' && $('#password').val()) {
```
Comment (bwv-hieu-nt) [inline resources/js/main.js:31]: Tham khảo jQuery validate submitHandler

File: resources/views/auth/login.blade.php [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+                    {{ $errors->all()[0] }}
+                </div>
+            </div>
+        @endif
+
+        <form action="{{ route('login') }}" method="post" id="loginForm">
+            @csrf
+            <div>
+                <label for="email" class="d-flex justify-content-end">Email</label>
+                <div class="group-input"><input type="email" name="email" id="email" value="{{ old('email') }}"></div>
```
Comment (bwv-hieu-nt) [inline resources/views/auth/login.blade.php:26]: Tìm hiểu component để viết các element input(text,radio,checkbox,....) , label, button và các xử lý khác, cho việc tái sử dụng

**Review by bwv-hieu-nt — COMMENTED** [review]
Ngoài ra cần tìm hiểu thêm repository pattern để viết các câu truy vấn tại đây, không viết query ở controller
Tạo các folder service để quản lý bussiness logic của các function như handleLogin là một ví dụ

File: app/Http/Controllers/AuthController.php (hunk 8 above)
Comment (nghia2742-briswell) [reply app/Http/Controllers/AuthController.php:43]: Trong phương thức Auth::logout(); đã có sẵn chức năng xóa các session.

File: app/Http/Controllers/AuthController.php (hunk 7 above)
Comment (nghia2742-briswell) [reply app/Http/Controllers/AuthController.php:31]: Em thấy trong docs validate như vậy mà không nhắc đến value list, hôm qua em có QA anh Thái thì ảnh cũng nói là làm theo docs.

File: resources/views/layouts/master.blade.php [hunk 13]
```diff
@@ -7,16 +7,16 @@
...
     </head>
     <body>
     
         @yield('content')
 
-        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
-        <script src="https://cdn.jsdelivr.net/jquery.validation/1.16.0/jquery.validate.min.js"></script>
-        <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/js/bootstrap.bundle.min.js" integrity="sha384-MrcW6ZMFYlzcLA8Nl+NtUVF0sA7MsXsP1UyJoMp4YLEuNSfAP+JcXn/tWtIaxVXM" crossorigin="anonymous"></script>
+        <script src="{{ asset('js/jquery/3.7.1/jquery.min.js') }}"></script>
+        <script src="{{ asset('js/bootstrap/bootstrap.bundle.min.js') }}"></script>
```
Comment (bwv-hieu-nt) [inline resources/views/layouts/master.blade.php:18]: bootstrap.min.js là được rồi em, file bundle chứa nhiều thứ khác không cần thiết

File: app/Utils/MessageUtils.php
Comment (bwv-hieu-nt) [inline app/Utils/MessageUtils.php:1]: đặt theo cấu trúc của Laravel nha em, MessageUtil.php, nhớ sửa tên class

File: app/Http/Requests/LoginRequest.php [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+     *
+     * @return array
+     */
+    public function messages(): array
+    {
+        $errorMsg010 = MessageUtils::getMessage('E010');
+        $errorMsg004 = MessageUtils::getMessage('E004');
+
+        return [
+            'email.required' => str_replace('{0}', 'Email', $errorMsg010),
```
Comment (bwv-hieu-nt) [inline app/Http/Requests/LoginRequest.php:42]: Em nên truyền 1 param là tên cột vào trong hàm getMessage nha, rồi xử lý bên trong hàm là str_replace gì đó
VD: getMessage('E001', [x1,x2]). 
Ví dụ nhiều hơn tham số cho {0},{1} thì em nên truyền 1 array luôn chứa các tham số và xử lý

Lỗi required phải là E001 nha em, check lại common sheet messages nha

File: app/View/Components/Input.php
Comment (bwv-hieu-nt) [inline app/View/Components/Input.php:1]: Viết dạng component class thì không vấn đề gì nha, nhưng em nên cân nhắc thêm về việc sử dụng anonymous component

File: public/js/jquery/jquery.validation/jquery.validate.min.js
Comment (bwv-hieu-nt) [inline public/js/jquery/jquery.validation/jquery.validate.min.js:1]: file này em nên bỏ file .min, và tái sử dụng custom nó luôn cũng đc, vì có các method, extend message ở đây và custom dễ dàng hơn. Còn nếu rối em có thể tách riêng ra cũng được, miễn là phải import file này thêm

File: resources/css/app.css
Comment (bwv-hieu-nt) [inline resources/css/app.css:1]: các css của riêng từng màn hình em cứ tách ra 1 file css giống js ở thư mục screen nha, mục đích là dễ quản lý

File: resources/views/screens/auth/login.blade.php [hunk 15]
```diff
@@ -0,0 +1,41 @@
+@extends('layouts.master')
```
Comment (bwv-hieu-nt) [inline resources/views/screens/auth/login.blade.php:1]: layouts master này anh nghĩ em tách ra cho login luôn nha, nếu layout login khác hoàn toàn với layout của các cụm screen như user,product,...

File: app/View/Components/Input.php
Comment (nghia2742-briswell) [reply app/View/Components/Input.php:1]: Dạ ok anh

File: resources/views/screens/auth/login.blade.php (hunk 15 above)
Comment (nghia2742-briswell) [reply resources/views/screens/auth/login.blade.php:1]: Cái này em có chia một master layout khác cho phần giao diện admin rồi ạ
